presto-main/src/main/java/io/prestosql/execution/resourcegroups/InternalResourceGroup.java

```diff
@@ -646,87 +646,116 @@ private void updateEligibility()
                 parent.get().eligibleSubGroups.remove(this);
                 lastStartMillis = 0;
             }
-            parent.get().updateEligibility();
+            parent.get().updateEligibilityOnPathToRoot();
         }
     }
 
     private void startInBackground(ManagedQueryExecution query)
     {
         checkState(Thread.holdsLock(root), "Must hold lock to start a query");
         synchronized (root) {
-            runningQueries.add(query);
+            runningQueries.put(query, new ResourceUsage(0, 0));
             InternalResourceGroup group = this;
             while (group.parent.isPresent()) {
                 group.parent.get().descendantRunningQueries++;
                 group.parent.get().dirtySubGroups.add(group);
                 group = group.parent.get();
             }
-            updateEligibility();
+            updateEligibilityOnPathToRoot();
             executor.execute(query::startWaitingForResources);
         }
     }
 
     private void queryFinished(ManagedQueryExecution query)
     {
         synchronized (root) {
-            if (!runningQueries.contains(query) && !queuedQueries.contains(query)) {
-                // Query has already been cleaned up
-                return;
-            }
-            // Only count the CPU time if the query succeeded, or the failure was the fault of the user
-            if (!query.getErrorCode().isPresent() || query.getErrorCode().get().getType() == USER_ERROR) {
-                InternalResourceGroup group = this;
-                while (group != null) {
-                    group.cpuUsageMillis = saturatedAdd(group.cpuUsageMillis, query.getTotalCpuTime().toMillis());
-                    group = group.parent.orElse(null);
-                }
-            }
-            if (runningQueries.contains(query)) {
-                runningQueries.remove(query);
+            // Query is queued
+            if (queuedQueries.contains(query)) {
+                queuedQueries.remove(query);
                 InternalResourceGroup group = this;
                 while (group.parent.isPresent()) {
-                    group.parent.get().descendantRunningQueries--;
+                    group.parent.get().descendantQueuedQueries--;
                     group = group.parent.get();
                 }
+                updateEligibilityOnPathToRoot();
+                return;
             }
-            else {
-                queuedQueries.remove(query);
+
+            ResourceUsage lastUsage = runningQueries.get(query);
+            // Query is running
+            if (lastUsage != null) {
+                // CPU is measured cumulatively (i.e. total CPU used until this moment by the query). Memory is measured
+                // instantaneously (how much memory the query is using at this moment). At query completion, memory usage
+                // drops to zero. CPU usage is extracted from query execution.
```

Comment:
Remove this part of the comment: `CPU usage is extracted from query execution`, it seems redundant.